src/lib/shadow-shim-helper-rs/src/rootedcell/cell.rs

```diff
@@ -0,0 +1,163 @@
+use super::{Root, Tag};
+use std::cell::UnsafeCell;
+use vasi::VirtualAddressSpaceIndependent;
+
+/// Analagous to [std::cell::Cell]. In particular like [std::cell::Cell], it
+/// doesn't perform any atomic operations internally, making it relatively
+/// inexpensive.
+///
+/// Unlike [std::cell::Cell], this type is [Send] and [Sync] if `T` is
+/// [Send]. This is safe because the owner is required to prove access to the
+/// associated [Root], which is `![Sync]`, to access.
+#[derive(Debug)]
+#[repr(C)]
+pub struct RootedCell<T>
+where
+    T: Copy,
+{
+    tag: Tag,
+    val: UnsafeCell<T>,
+}
+
+// SAFETY: RootedCell is VirtualAddressSpaceIndependent as long as T is.
+unsafe impl<T> VirtualAddressSpaceIndependent for RootedCell<T> where
+    T: VirtualAddressSpaceIndependent + Copy
+{
+}
+
+impl<T> RootedCell<T>
+where
+    T: Copy,
+{
+    /// Create a RootedCell associated with `root`.
+    pub fn new(root: &Root, val: T) -> Self {
+        Self {
+            tag: root.tag(),
+            val: UnsafeCell::new(val),
+        }
+    }
+
+    pub fn get(&self, root: &Root) -> T {
+        // Prove that the root is held for this tag.
+        assert_eq!(
+            root.tag, self.tag,
+            "Expected {:?} Got {:?}",
+            self.tag, root.tag
+        );
+
+        unsafe { *self.val.get() }
+    }
+
+    pub fn get_mut(&mut self) -> &mut T {
+        // Since we have the only reference to `self`, we don't need to check the root.
+        unsafe { &mut *self.val.get() }
+    }
+
+    pub fn set(&self, root: &Root, val: T) {
+        // Prove that the root is held for this tag.
+        assert_eq!(
+            root.tag, self.tag,
+            "Expected {:?} Got {:?}",
+            self.tag, root.tag
+        );
+
+        unsafe { *self.val.get() = val }
+    }
+
+    pub fn replace(&self, root: &Root, val: T) -> T {
+        // Prove that the root is held for this tag.
+        assert_eq!(
+            root.tag, self.tag,
+            "Expected {:?} Got {:?}",
+            self.tag, root.tag
+        );
+
+        let old = unsafe { *self.val.get() };
+        unsafe { *self.val.get() = val }
+        old
+    }
```

Comment:
I didn't think about it too hard, but maybe double-check that it behaves correctly when the methods of `RootedCell` are called re-entrantly (I'll pretend that's a word). For example, do the `UnsafeCell` accesses behave correctly when `replace` is called while you're in the middle of executing `set`? For example, if `T` contains a reference to the outer `RootedCell` and `T` is dropped inside of `set`, the drop impl will be able to call methods on the `RootedCell` while it's in the middle of executing `set`. An example for `Cell` would be:

```rust
use std::cell::Cell;
use std::rc::{Rc, Weak};

struct Foo(i32, Option<Weak<Cell<Foo>>>);

impl std::ops::Drop for Foo {
    fn drop(&mut self) {
        println!("Droping id: {}", self.0);

        if let Some(weak) = &self.1 {
            if let Some(rc) = weak.upgrade() {
                // we're in the middle of set(), but we can call replace() here
                let _x = rc.replace(Foo(2, None));

                // here we drop Foo(1, _)
            }
        }
    }
}

fn main() {
    let x = Rc::new_cyclic(|weak| Cell::new(Foo(0, Some(Weak::clone(&weak)))));

    // will drop Foo(0, _) and Foo(1, _) during the set()
    x.set(Foo(1, None));

    // here we drop Foo(2, _)
}
```

Reply:
I don't see a problem with doing something analogous here. I considered adding this as a test just to be sure, but implementing `new_cyclic` looks a bit tricky. It also looks potentially useful and I think would simplify the Process and (soon Thread) constructors but I don't want to dive down that rabbit hole at the moment :). I could work around not having it here, but that gets to be a fairly complex unit test; I'm not sure it's worth it.

Anyway, I think it's clearer where things get dropped etc after the other changes. Wdyt?

Reply:
Looks good to me!